File: haskell/pushVowels.hs
import Data.List
-- This function can push vowels to the end of a string 
pushVowels :: [Char] -> [Char]
-- pushVowels "nebraska" = "nbrskeaa"
-- pushVowels "nbc" = "nbc"
-- pushVowels "eue" = "eue"
-- pushVowels "oats" = "tsoa"
-- pushVowels "a" = "a"

-- base case(s)
pushVowels "" = ""
pushVowels [e] = [e]
-- general case
pushVowels (h:t)
 | (not (vowel h)) = h:(pushVowels t) -- don't push head down into tail
 | (not (vowel n)) = n:(pushVowels (h:r)) -- swap head and next, then push head and rest
 | (vowel h) && (vowel n) && (c == []) = h:t -- neighboring vowels, only vowels later
 | otherwise = pushVowels (h:(pushVowels t)) -- neighboring vowels, vowels later
 where n = head t
       r = tail t
       c = [x|x<-r,(not (vowel x))]

member :: Eq a => a -> [a] -> Bool
member _ [] = False -- nothing is a member of an empty list
member e (h:t) = e == h || (member e t) -- something is a member of a list if it is the head or a member of the tail

vowel :: Char -> Bool
vowel letter = (member letter "aeiou")

File: prolog/pushVowels.pl
vowel(a).
vowel(e).
vowel(i).
vowel(o).
vowel(u).

pushVowels([],[]):-!.
pushVowels([X],[X]):-!.
pushVowels([H,N|R],[H|Result]):-
  not(vowel(H)),
  pushVowels([N|R],Result),!.

pushVowels([H,N|R],[N|Result]):-
  not(vowel(N)),
  pushVowels([H|R],Result),!.

pushVowels([H,N|R],[H,N|R]):-
  vowel(H),
  vowel(N),
  findall(Letter,(member(Letter,R),not(vowel(Letter))),NotVowels),
  length(NotVowels,0),!.

pushVowels([H,N|R],Result):-
  vowel(H),
  vowel(N),
  findall(Letter,(member(Letter,R),not(vowel(Letter))),NotVowels),
  not(NotVowels = []),
  pushVowels([N|R],Re),
  pushVowels([H|Re], Result),!.

member(X,[X|_]).
member(X,[_|T]):-
  member(X,T).
